File: ios/Sources/ACTemplating/ExpressionEngine.swift
import Foundation

/// Protocol for expression engine implementations
/// Ported from production Teams-AdaptiveCards-Mobile SDK's protocol-oriented architecture.
/// Enables dependency injection and testability.
public protocol ExpressionEvaluating {
    /// Evaluate an expression string against a data context
    /// - Parameters:
    ///   - expression: The expression string (e.g., "name.first")
    ///   - context: The data context to evaluate against
    /// - Returns: The evaluated result
    func evaluate(expression: String, context: DataContext) throws -> Any?
}

/// Protocol for template engine implementations
public protocol TemplateProcessing {
    /// Process a template JSON string with data
    /// - Parameters:
    ///   - template: The template JSON string
    ///   - data: The data to bind into the template
    /// - Returns: The processed JSON string
    func process(template: String, with data: [String: Any]) throws -> String
}

/// Protocol for registering custom functions
public protocol FunctionRegistering {
    /// Register a custom function
    /// - Parameters:
    ///   - name: The function name
    ///   - function: The function implementation
    func registerFunction(name: String, function: ExpressionFunction)

    /// Check if a function is registered
    func hasFunction(name: String) -> Bool
}

/// Default implementation that ties together parser, evaluator, and cache
public final class ExpressionEngine: ExpressionEvaluating {
    private let parser: ExpressionParser
    private let cache: ExpressionCache?
    private var customFunctions: [String: ExpressionFunction] = [:]

    /// Create an expression engine with optional caching
    /// - Parameter cache: Expression cache (nil to disable caching)
    public init(cache: ExpressionCache? = ExpressionCache()) {
        self.parser = ExpressionParser()
        self.cache = cache
    }

    /// Evaluate an expression string against a data context
    public func evaluate(expression: String, context: DataContext) throws -> Any? {
        let parsed: Expression
        if let cache = cache {
            parsed = try cache.getOrParse(expression, using: parser)
        } else {
            parsed = try parser.parse(expression)
        }

        let evaluator = ExpressionEvaluator(context: context)
        return try evaluator.evaluate(parsed)
    }

    /// Cache statistics (nil if caching is disabled)
    public var cacheStats: (hits: Int, misses: Int, hitRate: Double, count: Int)? {
        guard let cache = cache else { return nil }
        return (cache.hits, cache.misses, cache.hitRate, cache.count)
    }

    /// Clear the expression cache
    public func clearCache() {
        cache?.clear()
    }
}

extension ExpressionEngine: FunctionRegistering {
    public func registerFunction(name: String, function: ExpressionFunction) {
        customFunctions[name] = function
    }

    public func hasFunction(name: String) -> Bool {
        customFunctions[name] != nil
    }
}

File: android/ac-templating/src/main/kotlin/com/microsoft/adaptivecards/templating/ExpressionEngine.kt
package com.microsoft.adaptivecards.templating

import com.microsoft.adaptivecards.templating.functions.*

/**
 * Interface for evaluating expressions.
 * Provides abstraction for testability and swappable implementations.
 */
interface ExpressionEvaluating {
    /**
     * Evaluate an expression string against a data context.
     * @param expression The expression string
     * @param context The data context
     * @return The evaluated result
     */
    fun evaluate(expression: String, context: DataContext): Any?
}

/**
 * Interface for processing templates with data binding.
 */
interface TemplateProcessing {
    /**
     * Process a template string, replacing `${...}` expressions with evaluated values.
     * @param template The template string
     * @param context The data context
     * @return The processed string
     */
    fun processTemplate(template: String, context: DataContext): String
}

/**
 * Interface for registering custom functions.
 */
interface FunctionRegistering {
    /**
     * Register a custom function.
     * @param name Function name
     * @param function Function implementation
     */
    fun registerFunction(name: String, function: ExpressionFunction)
}

/**
 * Unified expression engine that ties together parsing, evaluation, and caching.
 *
 * Ported from production Teams-AdaptiveCards-Mobile SDK's protocol-oriented architecture.
 * Provides a single entry point for expression evaluation with optional caching.
 *
 * @param cache Optional expression cache for performance optimization
 */
class ExpressionEngine(
    private val cache: ExpressionCache? = ExpressionCache()
) : ExpressionEvaluating, TemplateProcessing, FunctionRegistering {

    private val parser = ExpressionParser()
    private val customFunctions = mutableMapOf<String, ExpressionFunction>()

    override fun evaluate(expression: String, context: DataContext): Any? {
        val parsed = if (cache != null) {
            cache.getOrParse(expression, parser)
        } else {
            parser.parse(expression)
        }

        val evaluator = ExpressionEvaluator(context)
        return evaluator.evaluate(parsed)
    }

    override fun processTemplate(template: String, context: DataContext): String {
        val regex = Regex("""\$\{(.+?)\}""")
        return regex.replace(template) { match ->
            val expr = match.groupValues[1]
            try {
                val result = evaluate(expr, context)
                result?.toString() ?: ""
            } catch (_: Exception) {
                match.value // Return original on error
            }
        }
    }

    override fun registerFunction(name: String, function: ExpressionFunction) {
        customFunctions[name] = function
    }

    /** Clear the expression cache */
    fun clearCache() {
        cache?.clear()
    }

    /** Get cache statistics */
    fun cacheStats(): CacheStats? {
        val c = cache ?: return null
        return CacheStats(
            entries = c.count,
            hits = c.hits,
            misses = c.misses,
            hitRate = c.hitRate
        )
    }

    data class CacheStats(
        val entries: Int,
        val hits: Int,
        val misses: Int,
        val hitRate: Double
    )
}
